### ISSUE TITLE
[FEATURE] configurable coverage levels

### ISSUE THREAD

**Issue by esatterwhite** [tapjs/tapjs#979]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Have you read the `CONTRIBUTING` guide on posting issues, and `CODE_OF_CONDUCT`?

- [X] yes I read the things

### Description

In previous versions of tap, it was possible to specify a minimum level of coverage
```yaml
lines: 95
functions: 95
statements: 95
branches: 95
```

These options have all been removed, and it doesn't see like there is a way to replicate this behavior. There is `allow-incomplete-coverage`, but as far as I can tell, this will allow 0% coverage - there is no in between.

I see this in the docs:
> A module with 99% code coverage can be considered as two modules; a large one that is tested, and a smaller one with no tests at all. If code is worth testing, it's worth testing completely. For that reason, the tap runner instruments code using the built in V8 coverage API, and considers incomplete coverage to be a test failure. If code coverage is complete, no coverage report is generated. If it is incomplete, or if no coverage is generated at all, then a report is printed and the process exits with an error status code.

I understand the intention around this, and I generally agree with the sentiment, but this isn't really practical. I would like to be able to enforce a minimum standard of quality - and enforcing that through code coverage in automated CI is a great place to start. Fail a build if coverage drops below a certain level, But the all or nothing approach makes that a high barrier to entry.

For example, inheriting large codebases with no tests. Going from 0 to 100% coverage in a reasonable amount of time isn't practical for many companies. But getting to a reasonable level of test coverage, and enforcing a new minimum, while working to improve that minimum incrementally over time is.

### Example

```yaml
lines: 95
functions: 95
statements: 95
branches: 95
```

**Comment by isaacs** [comment]
I wonder if there isn't a better feature hiding here for that use case. One thing I never liked about the levels configs is just how arbitrary and manual they feel. 

Really, it seems like, if you're not at full coverage (or even if you are, I guess) then thing you want is to never *reduce* coverage.

What if, instead of setting levels explicitly, there was an option to track the current coverage level on each run, and fail if it ever goes below that limit?

So you'd run `tap --increase-coverage`, which would imply `--allow-incomplete-coverage` and write the current levels to the config if they're not there. If they are there, and the current levels are below, fail. If they're at or above, pass, and write the new values back.

**Comment by esatterwhite** [comment]
> So you'd run `tap --increase-coverage`, which would imply `--allow-incomplete-coverage` and write the current levels to the config if they're not there. If they are there, and the current levels are below, fail. If they're at or above, pass, and write the new values back.

Ya, I could get behind this. I think this is usually what people want when trying to enforce coverage levels. It goes up, not down. And this would remove the manual maintenance and human error from trying to keep up with it.

**Comment by esatterwhite** [comment]
I'm not too familiar with how tap manages configuration, but if those coverage values are stored in the existing tap config - that could be `.taprc`, `package.json` or one or more shared config packages. Which one gets updated?

If the intent is for this to not be manage by a human, maybe a separate top level file?

**Comment by isaacs** [comment]
It'd be the same logic as using `tap config set key=value`, ie, update .taprc if found, otherwise package.json if found, otherwise create a .taprc.

**Comment by julie-ng** [comment]
I agree, it's **_too strict_** 👎 I've turned it off for now and am considering switching to another testing framework.

I understand the good intentions. But I'm not sure I agree with the philosophy. I think it's better to _encourage_ good practices and allow for growth than to _force_ people to be perfect from day 1, which I'm willing to bet it has the opposite effect and people just skip coverage/tests. It's just disheartening for a runner to say 90% is not enough.

**Comment by isaacs** [comment]
Don't be disheartened! Be inspired to write better, more testable code, and better tests to cover it fully!

https://www.youtube.com/watch?v=hJ4XHQHhmw8

**Comment by esatterwhite** [comment]
I would love to see this make it in. To be honest, this is keeping us from updating tap, and at this point we're are considering migrating away from it.

**Comment by esatterwhite** [comment]
@isaacs Would a PR for this be taken into consideration?

**Comment by justintime4tea** [comment]
Without any changes to tap, you can use c8 to enforce a specific threshold:

- https://github.com/tapjs/tapjs/issues/877#issuecomment-1751266240
- https://github.com/tapjs/tapjs/issues/877#issuecomment-1752603217

**Comment by esatterwhite** [comment]
@justintime4tea 
> Without any changes to tap, you can use c8 to enforce a specific threshold:
> 
> * [Extract coverage reporting from C8 #877 (comment)](https://github.com/tapjs/tapjs/issues/877#issuecomment-1751266240)
> * [Extract coverage reporting from C8 #877 (comment)](https://github.com/tapjs/tapjs/issues/877#issuecomment-1752603217)

Sure, but tap is already using c8. Its a bit ham-fisted to have to also install c8 and run it manually when tap is doing this already. 
Its just not configurable

**Comment by justintime4tea** [comment]
@esatterwhite 

True! It would be nice not to have to do extra work.

Looks like @isaacs is open for a PR for this:
https://github.com/tapjs/tapjs/issues/877#issuecomment-1751266240

Here seems to be some relevant code path:
https://github.com/tapjs/tapjs/blob/532287f65d814c7cb7872bf84b3d86b17d900e43/src/run/src/report.ts#L204

**Comment by esatterwhite** [comment]
Just to further my gripe a bit. If you are using shared configs it makes it hard for people to tweak the tap config. Consumers of your config cannot just tweak tap config, they also now have to setup c8 and configure it.

**Comment by justintime4tea** [comment]
> Just to further my gripe a bit. If you are using shared configs it makes it hard for people to tweak the tap config. Consumers of your config cannot just tweak tap config, they also now have to setup c8 and configure it.

@esatterwhite - I just added a PR to "bring back" minimum coverage thresholds. This should make it so folks won't need to go through the extra work of setting up c8. Should help with shared configs as well?

**Comment by isaacs** [comment]
Landed in c48f531b
